### PR TITLE
feat: resolve relative datasample spec path

### PR DESCRIPTION
**Pull request by ThibaultFy**
## Related issue

See slack https://owkin.slack.com/archives/C01UWGW68N8/p1696943277910689 for context
closes FL-1243

## Summary

Resolve `path` and `paths` on datasample spec.

### CODE REVIEW

File: substra/sdk/schemas.py [hunk 1]
```diff
@@ -155,6 +155,23 @@ def exclusive_paths(cls, values):  # noqa: N805
             raise ValueError("'path' or 'paths' field must be set.")
         return values
 
+    @pydantic.model_validator(mode="before")
+    def resolve_paths(cls, values):  # noqa: N805
```
Comment (thbcmlowk) [inline substra/sdk/schemas.py:159]: No typing on `values`?

Comment (ThibaultFy) [reply substra/sdk/schemas.py:159]: I just copy paste it from abose. We might add it though, but it's `values` is provided by pydantic so it's not that bad no ? Wdyt ?

File: substra/sdk/schemas.py [hunk 2]
```diff
@@ -155,6 +155,23 @@ def exclusive_paths(cls, values):  # noqa: N805
...
+        """Resolve given path is relative."""
+        if "paths" in values:
+            paths = []
+            for path in values["paths"]:
+                path = pathlib.Path(path)
+                paths.append(path.resolve()) if not path.is_absolute() else paths.append(path)
```
Comment (guilhem-barthes) [inline substra/sdk/schemas.py:165]:
```suggestion
                paths.append(path.resolve())
```
Do we need to check `is_absolute()`  as`path.resolve()` will return itself if already a absolute path?

File: substra/sdk/schemas.py (hunk 1 above)
Comment (guilhem-barthes) [reply substra/sdk/schemas.py:159]: I would rather use `values: Any): -> Any` or create a custom type (e.g. `ValueType`) and use `values: ValueType): -> ValueType`

**Review by guilhem-barthes — COMMENTED** [review]
Thanks for your work!

File: substra/sdk/schemas.py [hunk 3]
```diff
@@ -147,14 +147,31 @@ def is_many(self):
         return self.paths and len(self.paths) > 0
 
     @pydantic.model_validator(mode="before")
-    def exclusive_paths(cls, values):  # noqa: N805
+    def exclusive_paths(cls, values: typing.Any) -> typing.Any:  # noqa: N805
```
Comment (guilhem-barthes) [inline substra/sdk/schemas.py:150]: can you remove the`# noqa: N805`?

File: substra/sdk/schemas.py [hunk 4]
```diff
@@ -147,14 +147,31 @@ def is_many(self):
...
         """Check that one and only one path(s) field is defined."""
         if "paths" in values and "path" in values:
             raise ValueError("'path' and 'paths' fields are exclusive.")
         if "paths" not in values and "path" not in values:
             raise ValueError("'path' or 'paths' field must be set.")
         return values
 
+    @pydantic.model_validator(mode="before")
+    def resolve_paths(cls, values: typing.Any) -> typing.Any:  # noqa: N805
```
Comment (guilhem-barthes) [inline substra/sdk/schemas.py:159]: can you remove the `# noqa: N805`? It should be handled by the typing

File: substra/sdk/schemas.py [hunk 5]
```diff
@@ -147,14 +147,31 @@ def is_many(self):
         return self.paths and len(self.paths) > 0
 
     @pydantic.model_validator(mode="before")
-    def exclusive_paths(cls, values):  # noqa: N805
+    def exclusive_paths(cls, values: typing.Any) -> typing.Any:  # noqa: N805
         """Check that one and only one path(s) field is defined."""
         if "paths" in values and "path" in values:
             raise ValueError("'path' and 'paths' fields are exclusive.")
         if "paths" not in values and "path" not in values:
             raise ValueError("'path' or 'paths' field must be set.")
```
Comment (guilhem-barthes) [inline substra/sdk/schemas.py:155]: Could they define both keys but with value `None`?

File: substra/sdk/schemas.py (hunk 3 above)
Comment (ThibaultFy) [reply substra/sdk/schemas.py:150]: Nop, its because it does not like that the first arg is not "self" :/

File: substra/sdk/schemas.py (hunk 5 above)
Comment (ThibaultFy) [reply substra/sdk/schemas.py:155]: I don't get what you're refering too. Could you provide an example plz ?

File: substra/sdk/schemas.py (hunk 3 above)
Comment (guilhem-barthes) [reply substra/sdk/schemas.py:150]: If this method entends to receive a class as first argument, pydantic recommends decorate with `@classmethod` for a proper type checking ([model_validator](https://docs.pydantic.dev/latest/concepts/validators/#model-validators))

Comment (ThibaultFy) [reply substra/sdk/schemas.py:150]: nice thx :)

File: substra/sdk/schemas.py (hunk 5 above)
Comment (guilhem-barthes) [reply substra/sdk/schemas.py:156]: It looks like we allow `None` as value on these fields, therefore if a user explicitely define only one of this key to `None` (e.g. `DataSampleSpec(path=None, data_manager_keys=[])`, the check `"paths" not in values and "path" not in values` will return False (as the key is present, but the value is `None`).

Comment (ThibaultFy) [reply substra/sdk/schemas.py:156]: Ok I see. I added some tests, and it appears that this case is dealts with in the resolve path, as `None` will raise a `TypeError`. This is not an explicit solution but I added the tests to ensure that this behavior is taken into account. What do you think ?

Comment (guilhem-barthes) [reply substra/sdk/schemas.py:156]: I don't know if `TypeError` is the clearest for the user, but maybe it is worth opening its own PR to review how we manage this edge case?

Comment (ThibaultFy) [reply substra/sdk/schemas.py:156]: Yep True... I rm the tests, and +1 to deal with this in a dedicated PR

**Review by guilhem-barthes — APPROVED** [review]
Thanks for your work!
